src/Symfony/Component/Translation/Tests/DependencyInjection/DataCollectorTranslatorPassTest.php

```diff
@@ -0,0 +1,121 @@
+        $this->container->setParameter('translator_not_implementing_bag', 'Symfony\Bundle\FrameworkBundle\Tests\DependencyInjection\Compiler\TranslatorWithTranslatorBag');
+
+        $this->container->register('translator.data_collector', DataCollectorTranslator::class)
+            ->setDecoratedService('translator')
+            ->setArguments([new Reference('translator.data_collector.inner')])
+        ;
+
+        $this->container->register('data_collector.translation', TranslationDataCollector::class)
+            ->setArguments([new Reference('translator.data_collector')])
+        ;
+    }
+
+    /**
+     * @dataProvider getImplementingTranslatorBagInterfaceTranslatorClassNames
+     */
+    public function testProcessKeepsDataCollectorTranslatorIfItImplementsTranslatorBagInterface($class)
+    {
+        $this->container->register('translator', $class);
+
+        $this->dataCollectorTranslatorPass->process($this->container);
+
+        $this->assertTrue($this->container->hasDefinition('translator.data_collector'));
+    }
+
+    /**
+     * @dataProvider getImplementingTranslatorBagInterfaceTranslatorClassNames
+     */
+    public function testProcessKeepsDataCollectorIfTranslatorImplementsTranslatorBagInterface($class)
+    {
+        $this->container->register('translator', $class);
+
+        $this->dataCollectorTranslatorPass->process($this->container);
+
+        $this->assertTrue($this->container->hasDefinition('data_collector.translation'));
+    }
+
+    public static function getImplementingTranslatorBagInterfaceTranslatorClassNames()
+    {
+        return [
+            [Translator::class],
+            ['%translator_implementing_bag%'],
+        ];
+    }
+
+    /**
+     * @dataProvider getNotImplementingTranslatorBagInterfaceTranslatorClassNames
+     */
+    public function testProcessRemovesDataCollectorTranslatorIfItDoesNotImplementTranslatorBagInterface($class)
+    {
+        $this->container->register('translator', $class);
+
+        $this->dataCollectorTranslatorPass->process($this->container);
+
+        $this->assertFalse($this->container->hasDefinition('translator.data_collector'));
+    }
+
+    /**
+     * @dataProvider getNotImplementingTranslatorBagInterfaceTranslatorClassNames
+     */
+    public function testProcessRemovesDataCollectorIfTranslatorDoesNotImplementTranslatorBagInterface($class)
+    {
+        $this->container->register('translator', $class);
+
+        $this->dataCollectorTranslatorPass->process($this->container);
+
+        $this->assertFalse($this->container->hasDefinition('data_collector.translation'));
+    }
+
+    public static function getNotImplementingTranslatorBagInterfaceTranslatorClassNames()
+    {
+        return [
+            ['Symfony\Bundle\FrameworkBundle\Tests\DependencyInjection\Compiler\TranslatorWithTranslatorBag'],
+            ['%translator_not_implementing_bag%'],
+        ];
+    }
+}
+
+class TranslatorWithTranslatorBag implements TranslatorInterface
```

Comment:
I suggest renaming that class as it is precisely about _not_ implementing TranslatorBagInterface.